web/api/v1/api.go

```diff
@@ -805,6 +808,93 @@ func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
 	return apiAlerts
 }
 
+type metricsWithMetadata map[string]metadataSet
+
+type metadataSet map[metadata]struct{}
+
+func (ms metadataSet) MarshalJSON() ([]byte, error) {
+	var res bytes.Buffer
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+	encoder := json.NewEncoder(&res)
+
+	res.WriteByte('[')
+	for metadata := range ms {
+		if err := encoder.Encode(metadata); err != nil {
+			return nil, err
+		}
+	}
+	res.WriteByte(']')
+
+	return res.Bytes(), nil
+}
+
+type metadata struct {
+	Type textparse.MetricType `json:"type"`
+	Help string               `json:"help"`
+	Unit string               `json:"unit"`
+}
+
+func (api *API) metricMetadata(r *http.Request) apiFuncResult {
+	var metrics metricsWithMetadata
+
+	limit := -1
+	if s := r.FormValue("limit"); s != "" {
+		var err error
+		if limit, err = strconv.Atoi(s); err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, errors.New("limit must be a number")}, nil, nil}
+		}
+	}
+
+	metrics = make(metricsWithMetadata)
+	for _, tt := range api.targetRetriever.TargetsActive() {
+		for _, t := range tt {
+
+		list:
+			for _, mm := range t.MetadataList() {
+				_, ok := metrics[mm.Metric]
+				m := metadata{Type: mm.Type, Help: mm.Help, Unit: mm.Unit}
+
+				// If the metric name is not in the map, we don't have any metadata yet.
+				if !ok {
+					metrics[mm.Metric] = make(map[metadata]struct{})
+					metrics[mm.Metric][m] = struct{}{}
+					continue
+				}
+
+				// If the metric name is in the map, find if the metadata is already in the set.
+				// This collapses the same metadata from different targets.
+				_, ok = metrics[mm.Metric][m]
+				if ok {
+					continue list
+				}
+
+				// By now, we know the metadata is not in the set - add it.
+				metrics[mm.Metric][m] = struct{}{}
+			}
+		}
+	}
+
+	if len(metrics) == 0 {
+		return apiFuncResult{nil, &apiError{errorNotFound, errors.New("No metadata available")}, nil, nil}
+	}
+
+	if limit >= 0 && len(metrics) >= limit {
+		lm := make(metricsWithMetadata, limit)
+
+		for k, v := range metrics {
```

Comment:
You can do `metrics[:limit]`

Reply:
Metrics is actually a map, if my understanding is correct - I can't slice a map?

Reply:
Ah, indeed.